In [4]:
import pandas as pd
import numpy as np

In [5]:
life=pd.read_csv('Life Expectancy Data.csv')

In [6]:
life.head()

Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   ...  \
0             62     0.01               71.279624         65.0      1154  ...   
1             64     0.01               73.523582         62.0       492  ...   
2             66     0.01               73.219243         64.0       430  ...   
3             69     0.01               78.184215         67.0      2787  ...   
4             71     0.01                7.097109         68.0      3013  ...   

   Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0    6.0               8.16         65.0        0.1  584.259210  33736494.0   
1   58.0               8.18         62.0        0.1  612.696514    327582.0   
2   62.0               8.13         64.0        0.1  631.744976  31731688.0   
3   67.0               8.52         67.0        0.1  669.959000   3696958.0   
4   68.0               7.87         68.0        0.1   63.537231   2978599.0   

    thinness  1-19 years   thinness 5-9 years  \
0                   17.2                 17.3   
1                   17.5                 17.5   
2                   17.7                 17.7   
3                   17.9                 18.0   
4                   18.2                 18.2   

   Income composition of resources  Schooling  
0                            0.479       10.1  
1                            0.476       10.0  
2                            0.470        9.9  
3                            0.463        9.8  
4                            0.454        9.5  

[5 rows x 22 columns]

In [7]:
life.shape

(2938, 22)

In [8]:
life.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10   BMI                             2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

## Cleaning Data Agendas:

- Proper = Country, Year, Status
- Life expectancy, Adult Mortality = 2928 non-null, 10 values are null
- Alcohol, HepatitisB, BMI, Polio, Total Expenditure, Diphttheria, GDP, Population, thinness1-19years, thinness5-9years, Incomecompositionofresources, Schooling.
- Above HepatitisB, Polio, Diphtheria, Population, AdultMortality = float -> int;
- 'nan' spotted in lifeexpectancy column
- check BMI values like 6.1, 1.0, 4.3 are these values are valid

In [9]:
life['Status'].unique()

array(['Developing', 'Developed'], dtype=object)

In [10]:
life['Population'].unique()

array([33736494.,   327582., 31731688., ...,   125525., 12366165.,
       12222251.])

In [11]:
life['Year'].unique()

array([2015, 2014, 2013, 2012, 2011, 2010, 2009, 2008, 2007, 2006, 2005,
       2004, 2003, 2002, 2001, 2000])

In [12]:
life.columns=life.columns.str.replace(' ','')

In [13]:
life

Country  Year      Status  Lifeexpectancy  AdultMortality  \
0     Afghanistan  2015  Developing            65.0           263.0   
1     Afghanistan  2014  Developing            59.9           271.0   
2     Afghanistan  2013  Developing            59.9           268.0   
3     Afghanistan  2012  Developing            59.5           272.0   
4     Afghanistan  2011  Developing            59.2           275.0   
...           ...   ...         ...             ...             ...   
2933     Zimbabwe  2004  Developing            44.3           723.0   
2934     Zimbabwe  2003  Developing            44.5           715.0   
2935     Zimbabwe  2002  Developing            44.8            73.0   
2936     Zimbabwe  2001  Developing            45.3           686.0   
2937     Zimbabwe  2000  Developing            46.0           665.0   

      infantdeaths  Alcohol  percentageexpenditure  HepatitisB  Measles  ...  \
0               62     0.01              71.279624        65.0     1154  ...   
1               64     0.01              73.523582        62.0      492  ...   
2               66     0.01              73.219243        64.0      430  ...   
3               69     0.01              78.184215        67.0     2787  ...   
4               71     0.01               7.097109        68.0     3013  ...   
...            ...      ...                    ...         ...      ...  ...   
2933            27     4.36               0.000000        68.0       31  ...   
2934            26     4.06               0.000000         7.0      998  ...   
2935            25     4.43               0.000000        73.0      304  ...   
2936            25     1.72               0.000000        76.0      529  ...   
2937            24     1.68               0.000000        79.0     1483  ...   

      Polio  Totalexpenditure  Diphtheria  HIV/AIDS         GDP  Population  \
0       6.0              8.16        65.0       0.1  584.259210  33736494.0   
1      58.0              8.18        62.0       0.1  612.696514    327582.0   
2      62.0              8.13        64.0       0.1  631.744976  31731688.0   
3      67.0              8.52        67.0       0.1  669.959000   3696958.0   
4      68.0              7.87        68.0       0.1   63.537231   2978599.0   
...     ...               ...         ...       ...         ...         ...   
2933   67.0              7.13        65.0      33.6  454.366654  12777511.0   
2934    7.0              6.52        68.0      36.7  453.351155  12633897.0   
2935   73.0              6.53        71.0      39.8   57.348340    125525.0   
2936   76.0              6.16        75.0      42.1  548.587312  12366165.0   
2937   78.0              7.10        78.0      43.5  547.358878  12222251.0   

      thinness1-19years  thinness5-9years  Incomecompositionofresources  \
0                  17.2              17.3                         0.479   
1                  17.5              17.5                         0.476   
2                  17.7              17.7                         0.470   
3                  17.9              18.0                         0.463   
4                  18.2              18.2                         0.454   
...                 ...               ...                           ...   
2933                9.4               9.4                         0.407   
2934                9.8               9.9                         0.418   
2935                1.2               1.3                         0.427   
2936                1.6               1.7                         0.427   
2937               11.0              11.2                         0.434   

      Schooling  
0          10.1  
1          10.0  
2           9.9  
3           9.8  
4           9.5  
...         ...  
2933        9.2  
2934        9.5  
2935       10.0  
2936        9.8  
2937        9.8  

[2938 rows x 22 columns]

In [14]:
life['Lifeexpectancy'].unique()

array([65. , 59.9, 59.5, 59.2, 58.8, 58.6, 58.1, 57.5, 57.3, 57. , 56.7,
       56.2, 55.3, 54.8, 77.8, 77.5, 77.2, 76.9, 76.6, 76.2, 76.1, 75.3,
       75.9, 74.2, 73.5, 73. , 72.8, 73.3, 73.6, 72.6, 75.6, 75.4, 75.1,
       74.9, 74.7, 74.4, 74.1, 73.8, 73.4, 72.9, 72.3, 71.7, 71.6, 71.4,
       71.3, 52.4, 51.7, 51.1, 56. , 51. , 49.6, 49.1, 48.7, 48.2, 47.7,
       47.4, 47.1, 46.8, 46.5, 45.7, 45.3, 76.4, 75.7, 75.2, 75. , 74.8,
       74.6, 74. , 76.3, 76. , 75.5, 73.9, 73.2, 72.7, 72. , 82.8, 82.7,
       82.5, 82.3, 82. , 81.9, 81.7, 81.3, 81.2, 81. , 86. , 83. , 79.9,
       79.5, 81.5, 81.4, 81.1, 88. , 84. , 79.8, 79.4, 79.3, 78.8, 78.7,
       78.6, 78.1, 72.5, 72.2, 71.9, 71.1, 78. , 69.2, 68.4, 67.8, 67.5,
       66.6, 74.5, 73.1, 76.8, 76.7, 76.5, 75.8, 71.8, 71. , 77. , 69.9,
       69.5, 69.1, 68.6, 68.2, 67.3, 66.8, 66.3, 65.8, 65.3, 73.7, 70. ,
       69.8, 68.9, 68.1, 67.7, 67.2, 68. , 89. , 87. , 80. , 78.9, 78.3,
       77.6, 69.4, 69.6, 69. , 68.7, 68.5, 68.3, 60

In [15]:
life['AdultMortality'].unique()

array([263., 271., 268., 272., 275., 279., 281., 287., 295., 291., 293.,
         3., 316., 321.,  74.,   8.,  84.,  86.,  88.,  91.,   1.,   9.,
        99.,  15.,  17.,  18.,  14.,  11.,  19., 112., 113., 116., 119.,
       123., 126., 129., 132., 136., 146., 145., 335., 348., 355., 358.,
       361., 365., 369., 371., 375., 381., 382., 386., 388., 391.,  44.,
        48.,  13., 131., 133., 134., 138., 142., 144., 147., 149., 151.,
       153., 154., 156., 118.,  12., 121., 127., 137., 128., 141.,  59.,
         6.,  61.,  63.,  64.,  66.,  67.,  69.,  71.,  73.,  75.,  78.,
        65.,  68.,   7.,  77.,  76.,  81.,  85.,  92.,  96., 125., 162.,
        16., 172., 167., 161., 168., 171., 174., 189., 192.,  79.,  83.,
        87., 135., 152., 155., 158., 164., 173.,  98., 111., 115., 117.,
       196., 199.,  23., 194., 232., 222., 226., 224., 243., 252., 247.,
       253., 262., 254.,  89.,  95., 175., 177., 188., 186., 184., 181.,
       191., 197.,  21., 249., 251., 259., 278., 28

In [16]:
life['infantdeaths'].unique()

array([  62,   64,   66,   69,   71,   74,   77,   80,   82,   84,   85,
         87,   88,    0,    1,   21,   20,   19,   67,   72,   75,   78,
         81,   90,   92,   94,   95,   96,   97,    8,    9,   10,   11,
         12,    5,    6,    7,   98,  104,  111,  118,  126,  135,  144,
        154,  164,  174,  185,  196,  207,  219,  231,   25,   13,   14,
         15,    2,   42,   44,   46,   49,   51,   54,   57,   61,   65,
         70,  103,   38,   39,   40,   41,   43,   45,   47,   48,   22,
         23,   24,   58,   59,   60,   63,   16,   17,   18,   27,   50,
         53,   56,  157,  171,  201,  215,  248,  266,  285,  307,  332,
        360,  391,  422,  457,  490,  236,  237,  238,  239,  235,  233,
        232,  229,  227,  226,   55,    3,    4,  136,  140,  145,  150,
        156,  162,  169,  177,  186,  206,  217,  228,  246,  253,   37,
         26,   28,   29,   30,   31,   32,   33,   34,   35,   36,  910,
        957, 1000, 1100, 1200, 1300, 1400, 1500, 16

In [17]:
life['Measles'].unique()

array([  1154,    492,    430,   2787,   3013,   1989,   2861,   1599,
         1141,   1990,   1296,    466,    798,   2486,   8762,   6532,
            0,      9,     28,     10,     22,     68,      6,      7,
            8,     16,     18,    662,     63,     25,    112,    103,
          107,    217,    944,   2302,   3289,  15374,   5862,   2686,
          118,  11699,   8523,   4458,   1449,   1190,   2807,    265,
         1014,    765,    258,     29,   1196,  11945,   9046,   2219,
            1,      2,      3,     17,     33,     13,    137,   2281,
         1783,      4,     40,     69,     15,     74,    340,    158,
          199,    190,     70,    104,     65,     11,     91,     32,
          141,    108,    309,    117,     36,     52,     49,    448,
           20,     23,     90,    164,      5,    222,   1238,    827,
         1978,   4353,    574,    210,     46,     12,    240,    289,
          237,   1986,   5625,    788,    718,   2660,   2924,   6192,
      

In [18]:
life['under-fivedeaths'].unique()

array([  83,   86,   89,   93,   97,  102,  106,  110,  113,  116,  118,
        120,  122,    0,    1,   24,   23,   22,   25,   98,  101,  105,
        115,  121,  127,  133,  138,  143,  148,  152,  155,  157,  159,
        160,    9,   10,   11,   12,   13,   14,    2,    6,    7,    8,
        130,  139,  150,  161,  173,  186,  201,  215,  231,  247,  264,
        280,  298,  316,   39,   40,   15,   16,   17,   18,   19,   20,
          3,    4,   47,   49,   52,   54,   57,   61,   64,   68,   73,
         79,   85,  100,  109,   63,   66,   71,   74,   77,   80,   91,
         92,   31,   32,   33,   34,   35,   36,   37,   38,   41,   81,
         82,   84,   87,   90,   94,   95,   21,   26,   29,   76,   99,
         78,   75,  183,  198,  216,  233,  251,  268,  288,  308,  332,
        360,  392,  427,  467,  511,  558,  608,  312,  314,  317,  319,
        321,  323,  324,  326,  327,  329,  330,  331,   60,   62,   65,
         70,   72,    5,  194,  202,  211,  221,  2

In [19]:
life['Polio'].unique()

array([ 6., 58., 62., 67., 68., 66., 63., 64.,  5., 41., 36., 35., 24.,
       99., 98., 97., 95., 94., 92., 88., 86., 87., 89.,  7., 75., 73.,
       81., 65., 39.,  4., 37.,  3., 96., 93., 91., 85.,  9., 83., 84.,
       82., 71.,  8., 79., 78., 77., 53., 74., 76., 69., 45., 72., 54.,
       59., 61., 57., 47., 23., 46., 44., 42., 38., 52., 51., 43., 32.,
       49., 31., 26., 17., 55., 48., 56., nan, 33.])

In [20]:
life['Totalexpenditure'].unique()

array([ 8.16,  8.18,  8.13,  8.52,  7.87,  9.2 ,  9.42,  8.33,  6.73,
        7.43,  8.7 ,  8.79,  8.82,  7.76,  7.8 ,  8.2 ,  6.  ,  5.88,
        5.66,  5.59,  5.71,  5.34,  5.79,  5.87,  6.1 ,  5.86,  6.12,
        6.38,  6.27,  6.3 ,  6.26,   nan,  7.21,  7.12,  6.14,  5.29,
        5.12,  5.36,  4.2 ,  3.82,  3.36,  3.24,  3.54,  3.6 ,  3.73,
        3.84,  3.49,  3.31,  4.26,  3.3 ,  3.38,  3.39,  4.37,  4.54,
        4.1 ,  4.71,  4.41,  3.63,  5.38,  2.79,  5.54,  5.33,  5.39,
        5.65,  5.63,  4.86,  4.69,  4.27,  4.34,  4.21,  4.53,  4.48,
        4.13,  4.79,  4.99,  5.2 ,  5.89,  6.55,  7.63,  6.66,  6.49,
        6.68,  6.85,  6.84,  8.22,  8.31,  9.38,  9.21,  4.55,  3.71,
        4.56,  3.8 ,  4.31,  4.58,  5.25,  5.5 ,  5.56,  5.4 ,  5.94,
        6.25,  9.36,  9.5 ,  8.78,  8.53,  8.49,  8.45,  8.57,  8.32,
        8.39,  8.8 , 11.21, 11.14, 11.17,  1.94, 11.19,  1.6 ,  1.4 ,
        1.35,  1.53,  1.56,  1.48,  1.27,  1.12,  6.4 ,  5.37,  5.1 ,
        5.85,  6.17,

In [21]:
life['Diphtheria'].unique()

array([65., 62., 64., 67., 68., 66., 63., 58.,  5., 41., 36., 33., 24.,
       99., 98., 97., 95., 93., 88., 86., 87., 89., 77., 75., 71.,  6.,
       69., 73., 34., 38.,  4., 28., 94., 91., 96., 83.,  9., 92., 85.,
       84., 81., 79., 78., 76., 82.,  8., 74.,  7., 72., 45., 61., 54.,
       59., 47., 23., 42., 48., 51., 44., 37., 46., 39., 19., 25., 21.,
       26., 31., 56.,  3., 53., 16.,  2., 43., 35., 32., 57., 49., 55.,
       52., nan, 29., 27.])

In [22]:
life['HIV/AIDS'].unique()

array([ 0.1,  1.9,  2. ,  2.3,  2.6,  2.5,  2.4,  2.1,  0.2,  0.4,  0.8,
        0.7,  0.9,  0.5,  0.3,  0.6,  1.5,  1. ,  1.1,  1.2,  1.3,  1.4,
        1.6,  1.8,  2.2,  2.8,  4.4,  5.5,  6.2,  9. , 12.7, 13.4, 14.4,
       20.6, 28.4, 31.9, 34.6, 37.2, 38.8,  1.7,  2.9,  3.2,  3.6,  4. ,
        3.4,  3.8,  4.3,  4.8,  5.1,  5.2,  5.3,  3.3,  3.7,  4.1,  5.8,
        6.1,  6.5,  6.7,  6.9,  7. ,  7.1,  3.5,  4.6,  4.9,  6.3,  7.2,
        7.4,  7.5,  7.7,  7.8,  7.9,  4.5,  6.6,  7.3,  8.3, 10. , 11.2,
       12. , 12.8, 13.9, 14.3,  3.1,  3.9,  4.7,  5. ,  3. ,  5.9,  6.8,
        7.6,  2.7,  4.2,  5.7,  8.2,  6. ,  8. ,  8.9, 10.7, 10.9, 11.1,
       11. , 10.1,  9.5,  5.6,  5.4,  6.4,  9.1, 10.3, 11.6, 13.2, 14.5,
       15.7, 16.9, 18.1,  9.3,  9.4,  9.6, 10.5, 18.2, 27.3, 30. , 34.1,
       34.8, 33.8, 32.5, 31.2, 29.8, 13.7, 14.9, 19.3, 21.1, 22.4, 23.4,
       24.2, 24.7, 25.1, 25.5, 10.8, 11.3, 12.6, 14.1, 16.3, 16.2, 15.9,
       15.3, 12.2,  8.7, 11.7, 15.2, 19.2, 22.1, 24

In [23]:
life['GDP'].unique()

array([584.25921  , 612.696514 , 631.744976 , ...,  57.34834  ,
       548.587312 , 547.3588785])

In [24]:
life['Population'].unique()

array([33736494.,   327582., 31731688., ...,   125525., 12366165.,
       12222251.])

In [25]:
life['thinness1-19years'].unique()

array([17.2, 17.5, 17.7, 17.9, 18.2, 18.4, 18.6, 18.8, 19. , 19.2, 19.3,
       19.5, 19.7, 19.9,  2.1,  2.3,  1.2,  1.3,  1.4,  1.5,  1.6,  1.7,
        1.8,  1.9,  2. ,  6. ,  5.9,  6.1,  6.2,  6.3,  6.4,  6.5,  8.3,
        8.5,  8.6,  8.8,  8.9,  9.1,  9.3,  9.5,  9.6,  9.8,  1. ,  3.3,
        3.4,  3.5,  3.6,  3.7,  1.1,  0.6,  0.7,  2.8,  2.9,  3. ,  3.1,
        3.2,  2.5,  2.6,  2.7, 18.1, 18.3, 18.5, 18.7, 18.9, 19.1,  3.8,
        3.9,  4. ,  4.1,  4.2,  4.3,  2.2,  2.4,  0.9,  0.8,  6.9,  7.1,
        7.2,  7.4,  7.6,  7.8,  7.9,  8.1,  8.7,  9.7, 15.4, 15.7, 15.9,
       16.1, 16.3, 16.6, 16.8, 17.1, 17.3, 17.8, 18. ,  6.7,  7. ,  7.3,
        7.7,  8. ,  8.4,  9.2, 11.4, 11.8, 12.3,  5.7,  5.8,  6.6,  6.8,
        8.2,  9. , 11. , 11.2,  7.5,  5.5,  5.6,  9.4, 11.1, 11.3, 11.5,
       11.6,  0.5,  9.9,  4.4,  4.5,  4.7,  4.8,  5. ,  5.1,  5.3,  0.1,
        4.9,  5.2,  5.4, 11.7, 11.9, 12.1, 12.2, 12.4,  4.6, 26.7, 26.8,
       26.9, 27. , 27.1, 27.2, 27.3, 27.4, 27.5, 27

In [26]:
life['thinness5-9years'].unique()

array([17.3, 17.5, 17.7, 18. , 18.2, 18.4, 18.7, 18.9, 19.1, 19.3, 19.5,
       19.7, 19.9,  2.2,  2.4,  2.5,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,
        1.9,  2. ,  2.1,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  8.2,
        8.3,  8.5,  8.6,  8.8,  9. ,  9.2,  9.4,  9.6,  9.7,  9.9,  1.1,
        3.3,  3.4,  3.5,  3.6,  0.9,  1. ,  0.6,  0.7,  2.9,  2.8,  3. ,
        3.1,  2.6, 18.3, 18.6, 18.8, 19. , 19.2, 19.4,  2.3,  2.7, 21.1,
       21.3, 21.5,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  0.8,  6.8,  6.9,
        7.1,  7.3,  7.5,  7.6,  7.8,  8. ,  8.4, 16. , 16.2, 16.5, 16.7,
       17. , 17.8, 18.1, 19.6,  1.2,  3.2,  6.7,  7. ,  7.4, 11.3, 11.8,
       12.2,  5.1,  5.2,  5.3,  5.4,  5.5,  5.6,  5.7,  7.7,  7.9,  8.1,
        9.1,  9.3,  9.8, 11.1,  7.2,  8.7,  8.9,  6.6,  9.5, 11. , 11.2,
       11.4, 11.5, 11.6, 11.7, 11.9,  6.5,  0.5,  0.4,  4.3,  4.4,  4.6,
        4.7,  4.9,  5. ,  0.1, 12.1, 12.3,  4.5,  4.8, 27.3, 27.4, 27.5,
       27.6, 27.7, 27.8, 27.9, 28. , 28.1, 28.2, 28

In [27]:
life['Incomecompositionofresources'].unique()

array([0.479, 0.476, 0.47 , 0.463, 0.454, 0.448, 0.434, 0.433, 0.415,
       0.405, 0.396, 0.381, 0.373, 0.341, 0.34 , 0.338, 0.762, 0.761,
       0.759, 0.752, 0.738, 0.725, 0.721, 0.713, 0.703, 0.696, 0.685,
       0.681, 0.674, 0.67 , 0.662, 0.656, 0.743, 0.741, 0.737, 0.732,
       0.724, 0.714, 0.705, 0.697, 0.69 , 0.686, 0.68 , 0.673, 0.663,
       0.653, 0.644, 0.636, 0.531, 0.527, 0.523, 0.508, 0.495, 0.488,
       0.48 , 0.468, 0.439, 0.426, 0.406, 0.401, 0.391, 0.382, 0.784,
       0.782, 0.781, 0.778, 0.783, 0.788, 0.786, 0.773, 0.   , 0.826,
       0.825, 0.823, 0.822, 0.816, 0.802, 0.794, 0.792, 0.78 , 0.775,
       0.77 , 0.776, 0.771, 0.764, 0.739, 0.736, 0.729, 0.72 , 0.707,
       0.692, 0.679, 0.668, 0.657, 0.645, 0.639, 0.937, 0.936, 0.933,
       0.93 , 0.927, 0.925, 0.921, 0.918, 0.915, 0.91 , 0.908, 0.905,
       0.902, 0.899, 0.897, 0.892, 0.887, 0.884, 0.88 , 0.872, 0.87 ,
       0.864, 0.86 , 0.854, 0.848, 0.841, 0.837, 0.847, 0.833, 0.758,
       0.745, 0.742,

In [28]:
life['Schooling'].unique()

array([10.1, 10. ,  9.9,  9.8,  9.5,  9.2,  8.9,  8.7,  8.4,  8.1,  7.9,
        6.8,  6.5,  6.2,  5.9,  5.5, 14.2, 13.3, 12.5, 12.2, 12. , 11.6,
       11.4, 10.8, 10.9, 10.7, 10.6, 14.4, 14. , 13.6, 13.1, 12.6, 12.3,
       11.7, 11.5, 11.1, 10.3,  9.4,  9. ,  8.5,  7.7,  7.2,  6.4,  5.1,
        4.6, 13.9, 13.8, 14.1, 14.5, 14.7,  0. , 17.3, 17.2, 17.1, 16.8,
       16.5, 16.3, 16.1, 16.4, 15.6, 15. , 12.7, 11.9, 11.2, 20.4, 20.3,
       20.1, 19.8, 19.5, 19.1, 19. , 20.7, 20.6, 20.5, 15.9, 15.7, 15.4,
       15.3, 15.1, 15.2, 14.9, 15.5, 11.8, 11. , 10.4, 12.4, 12.1, 13.7,
       13.5, 13.2, 10.2,  8.6,  8.2,  7.5,  7.3, 15.8, 14.8, 14.6, 16.6,
       16.2, 18.8, 18.6, 18.2, 18. , 12.8, 12.9,  9.3,  9.1,  6.6, 10.5,
        9.6,  8.8,  8. ,  7.6, 14.3, 13.4,  6.7,  6.3,  5.4,  4.9,  4.7,
        4.3,  3.9,  3.8,  3.6,  3.5,  3.4,  5.6,  5.2,  4.4,  4.5,  nan,
       11.3,  9.7,  8.3,  7.1,  6.9,  6. ,  5.7,  5.3, 13. , 17.6, 17.7,
       16. , 19.2, 18.7, 18.4, 16.9, 16.7,  5. ,  4

In [29]:
life.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Country                       2938 non-null   object 
 1   Year                          2938 non-null   int64  
 2   Status                        2938 non-null   object 
 3   Lifeexpectancy                2928 non-null   float64
 4   AdultMortality                2928 non-null   float64
 5   infantdeaths                  2938 non-null   int64  
 6   Alcohol                       2744 non-null   float64
 7   percentageexpenditure         2938 non-null   float64
 8   HepatitisB                    2385 non-null   float64
 9   Measles                       2938 non-null   int64  
 10  BMI                           2904 non-null   float64
 11  under-fivedeaths              2938 non-null   int64  
 12  Polio                         2919 non-null   float64
 13  Tot

## Cleaning Data Agendas:

- Proper = Country, Year, Status
- Life expectancy, Adult Mortality = 2928 non-null, 10 values are null
- Alcohol, HepatitisB, BMI, Polio, Total Expenditure, Diphttheria, GDP, Population, thinness1-19years, thinness5-9years, Incomecompositionofresources, Schooling.
- Above HepatitisB, Polio, Diphtheria, Population, AdultMortality = float -> int;
- 'nan' spotted in lifeexpectancy column
- check BMI values like 6.1, 1.0, 4.3 are these values are valid

## Cleaning:

In [30]:
backup=life.copy()

## Replacing NULL Values with mean values

In [31]:
mean_value=life['Lifeexpectancy'].mean()
print(mean_value)

69.22493169398908


In [32]:
life['Lifeexpectancy'].fillna(value=mean_value, inplace=True)

In [33]:
mean_value=life['AdultMortality'].mean()
print(mean_value)

164.79644808743168


In [34]:
life['AdultMortality'].fillna(value=mean_value, inplace=True)

In [35]:
life['AdultMortality'].astype(int)

0       263
1       271
2       268
3       272
4       275
       ... 
2933    723
2934    715
2935     73
2936    686
2937    665
Name: AdultMortality, Length: 2938, dtype: int64

In [36]:
mean_value=life['Alcohol'].mean()
print(mean_value)
life['Alcohol'].fillna(value=mean_value, inplace=True)

4.602860787172012


In [37]:
mean_value=life['HepatitisB'].mean()
print(mean_value)
life['HepatitisB'].fillna(value=mean_value, inplace=True)

80.94046121593291


In [38]:
mean_value=life['BMI'].mean()
print(mean_value)
life['BMI'].fillna(value=mean_value, inplace=True)

38.321246556473824


In [39]:
mean_value=life['Polio'].mean()
print(mean_value)
life['Polio'].fillna(value=mean_value, inplace=True)

82.55018842069202


In [40]:
mean_value=life['Totalexpenditure'].mean()
print(mean_value)
life['Totalexpenditure'].fillna(value=mean_value, inplace=True)

5.9381895280235995


In [41]:
mean_value=life['Diphtheria'].mean()
print(mean_value)
life['Diphtheria'].fillna(value=mean_value, inplace=True)

82.32408359027065


In [42]:
mean_value=life['GDP'].mean()
print(mean_value)
life['GDP'].fillna(value=mean_value, inplace=True)

7483.158469138474


In [43]:
mean_value=life['Population'].mean()
print(mean_value)
life['Population'].fillna(value=mean_value, inplace=True)

12753375.120052494


In [44]:
mean_value=life['thinness1-19years'].mean()
print(mean_value)
life['thinness1-19years'].fillna(value=mean_value, inplace=True)

4.839703856749312


In [45]:
mean_value=life['thinness5-9years'].mean()
print(mean_value)
life['thinness5-9years'].fillna(value=mean_value, inplace=True)

4.870316804407714


In [46]:
mean_value=life['Incomecompositionofresources'].mean()
print(mean_value)
life['Incomecompositionofresources'].fillna(value=mean_value, inplace=True)

0.6275510645976182


In [47]:
mean_value=life['Schooling'].mean()
print(mean_value)
life['Schooling'].fillna(value=mean_value, inplace=True)

11.992792792792793


# Converting features into their original datatype - HepatitisB, Polio, Diphtheria, GDP, Population, AdultMortality

In [48]:
life['HepatitisB']=life['HepatitisB'].astype(int)

In [49]:
life['Polio']=life['Polio'].astype(int)

In [50]:
life['Diphtheria']=life['Diphtheria'].astype(int)

In [51]:
life['Population']=life['Population'].astype(int)

In [52]:
life['AdultMortality']=life['AdultMortality'].astype(int)

In [53]:
life['GDP']=life['AdultMortality'].astype(int)

In [54]:
life.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Country                       2938 non-null   object 
 1   Year                          2938 non-null   int64  
 2   Status                        2938 non-null   object 
 3   Lifeexpectancy                2938 non-null   float64
 4   AdultMortality                2938 non-null   int64  
 5   infantdeaths                  2938 non-null   int64  
 6   Alcohol                       2938 non-null   float64
 7   percentageexpenditure         2938 non-null   float64
 8   HepatitisB                    2938 non-null   int64  
 9   Measles                       2938 non-null   int64  
 10  BMI                           2938 non-null   float64
 11  under-fivedeaths              2938 non-null   int64  
 12  Polio                         2938 non-null   int64  
 13  Tot

In [55]:
life

Country  Year      Status  Lifeexpectancy  AdultMortality  \
0     Afghanistan  2015  Developing            65.0             263   
1     Afghanistan  2014  Developing            59.9             271   
2     Afghanistan  2013  Developing            59.9             268   
3     Afghanistan  2012  Developing            59.5             272   
4     Afghanistan  2011  Developing            59.2             275   
...           ...   ...         ...             ...             ...   
2933     Zimbabwe  2004  Developing            44.3             723   
2934     Zimbabwe  2003  Developing            44.5             715   
2935     Zimbabwe  2002  Developing            44.8              73   
2936     Zimbabwe  2001  Developing            45.3             686   
2937     Zimbabwe  2000  Developing            46.0             665   

      infantdeaths  Alcohol  percentageexpenditure  HepatitisB  Measles  ...  \
0               62     0.01              71.279624          65     1154  ...   
1               64     0.01              73.523582          62      492  ...   
2               66     0.01              73.219243          64      430  ...   
3               69     0.01              78.184215          67     2787  ...   
4               71     0.01               7.097109          68     3013  ...   
...            ...      ...                    ...         ...      ...  ...   
2933            27     4.36               0.000000          68       31  ...   
2934            26     4.06               0.000000           7      998  ...   
2935            25     4.43               0.000000          73      304  ...   
2936            25     1.72               0.000000          76      529  ...   
2937            24     1.68               0.000000          79     1483  ...   

      Polio  Totalexpenditure  Diphtheria  HIV/AIDS  GDP  Population  \
0         6              8.16          65       0.1  263    33736494   
1        58              8.18          62       0.1  271      327582   
2        62              8.13          64       0.1  268    31731688   
3        67              8.52          67       0.1  272     3696958   
4        68              7.87          68       0.1  275     2978599   
...     ...               ...         ...       ...  ...         ...   
2933     67              7.13          65      33.6  723    12777511   
2934      7              6.52          68      36.7  715    12633897   
2935     73              6.53          71      39.8   73      125525   
2936     76              6.16          75      42.1  686    12366165   
2937     78              7.10          78      43.5  665    12222251   

      thinness1-19years  thinness5-9years  Incomecompositionofresources  \
0                  17.2              17.3                         0.479   
1                  17.5              17.5                         0.476   
2                  17.7              17.7                         0.470   
3                  17.9              18.0                         0.463   
4                  18.2              18.2                         0.454   
...                 ...               ...                           ...   
2933                9.4               9.4                         0.407   
2934                9.8               9.9                         0.418   
2935                1.2               1.3                         0.427   
2936                1.6               1.7                         0.427   
2937               11.0              11.2                         0.434   

      Schooling  
0          10.1  
1          10.0  
2           9.9  
3           9.8  
4           9.5  
...         ...  
2933        9.2  
2934        9.5  
2935       10.0  
2936        9.8  
2937        9.8  

[2938 rows x 22 columns]

In [56]:
life.describe()

Year  Lifeexpectancy  AdultMortality  infantdeaths      Alcohol  \
count  2938.000000     2938.000000     2938.000000   2938.000000  2938.000000   
mean   2007.518720       69.224932      164.793737     30.303948     4.602861   
std       4.613841        9.507640      124.080310    117.926501     3.916288   
min    2000.000000       36.300000        1.000000      0.000000     0.010000   
25%    2004.000000       63.200000       74.000000      0.000000     1.092500   
50%    2008.000000       72.000000      144.000000      3.000000     4.160000   
75%    2012.000000       75.600000      227.000000     22.000000     7.390000   
max    2015.000000       89.000000      723.000000   1800.000000    17.870000   

       percentageexpenditure   HepatitisB        Measles          BMI  \
count            2938.000000  2938.000000    2938.000000  2938.000000   
mean              738.251295    80.763445    2419.592240    38.321247   
std              1987.914858    22.589847   11467.272489    19.927677   
min                 0.000000     1.000000       0.000000     1.000000   
25%                 4.685343    80.000000       0.000000    19.400000   
50%                64.912906    87.000000      17.000000    43.000000   
75%               441.534144    96.000000     360.250000    56.100000   
max             19479.911610    99.000000  212183.000000    87.300000   

       under-fivedeaths        Polio  Totalexpenditure   Diphtheria  \
count       2938.000000  2938.000000       2938.000000  2938.000000   
mean          42.035739    82.546630          5.938190    82.321988   
std          160.445548    23.352184          2.400274    23.640087   
min            0.000000     3.000000          0.370000     2.000000   
25%            0.000000    78.000000          4.370000    78.000000   
50%            4.000000    93.000000          5.938190    93.000000   
75%           28.000000    97.000000          7.330000    97.000000   
max         2500.000000    99.000000         17.600000    99.000000   

          HIV/AIDS          GDP    Population  thinness1-19years  \
count  2938.000000  2938.000000  2.938000e+03        2938.000000   
mean      1.742103   164.793737  1.275338e+07           4.839704   
std       5.077785   124.080310  5.381546e+07           4.394535   
min       0.100000     1.000000  3.400000e+01           0.100000   
25%       0.100000    74.000000  4.189172e+05           1.600000   
50%       0.100000   144.000000  3.675929e+06           3.400000   
75%       0.800000   227.000000  1.275338e+07           7.100000   
max      50.600000   723.000000  1.293859e+09          27.700000   

       thinness5-9years  Incomecompositionofresources    Schooling  
count       2938.000000                   2938.000000  2938.000000  
mean           4.870317                      0.627551    11.992793  
std            4.482708                      0.204820     3.264381  
min            0.100000                      0.000000     0.000000  
25%            1.600000                      0.504250    10.300000  
50%            3.400000                      0.662000    12.100000  
75%            7.200000                      0.772000    14.100000  
max           28.600000                      0.948000    20.700000

In [57]:
life[life['Measles']==0]

Country  Year      Status  Lifeexpectancy  AdultMortality  \
16     Albania  2015  Developing            77.8              74   
17     Albania  2014  Developing            77.5               8   
18     Albania  2013  Developing            77.2              84   
22     Albania  2009  Developing            76.1              91   
23     Albania  2008  Developing            75.3               1   
...        ...   ...         ...             ...             ...   
2923  Zimbabwe  2014  Developing            59.2             371   
2924  Zimbabwe  2013  Developing            58.0             399   
2925  Zimbabwe  2012  Developing            56.6             429   
2926  Zimbabwe  2011  Developing            54.9             464   
2929  Zimbabwe  2008  Developing            48.2             632   

      infantdeaths  Alcohol  percentageexpenditure  HepatitisB  Measles  ...  \
16               0     4.60             364.975229          99        0  ...   
17               0     4.51             428.749067          98        0  ...   
18               0     4.76             430.876979          99        0  ...   
22               1     5.79             348.055952          98        0  ...   
23               1     5.61              36.622068          99        0  ...   
...            ...      ...                    ...         ...      ...  ...   
2923            23     6.50              10.822595          91        0  ...   
2924            25     6.39              10.666707          95        0  ...   
2925            26     6.09              92.602336          97        0  ...   
2926            28     6.00              63.750530          94        0  ...   
2929            30     3.56              20.843429          75        0  ...   

      Polio  Totalexpenditure  Diphtheria  HIV/AIDS  GDP  Population  \
16       99              6.00          99       0.1   74       28873   
17       98              5.88          98       0.1    8      288914   
18       99              5.66          99       0.1   84      289592   
22       98              5.79          98       0.1   91     2927519   
23       99              5.87          99       0.1    1     2947314   
...     ...               ...         ...       ...  ...         ...   
2923     92              6.44          91       6.3  371    15411675   
2924     95              6.88          95       6.8  399      155456   
2925     95              6.69          95       8.8  429     1471826   
2926     93              6.31          93      13.3  464    14386649   
2929     75              4.96          75      20.5  632    13558469   

      thinness1-19years  thinness5-9years  Incomecompositionofresources  \
16                  1.2               1.3                         0.762   
17                  1.2               1.3                         0.761   
18                  1.3               1.4                         0.759   
22                  1.5               1.6                         0.721   
23                  1.6               1.6                         0.713   
...                 ...               ...                           ...   
2923                5.9               5.7                         0.498   
2924                6.2               6.0                         0.488   
2925                6.5               6.4                         0.464   
2926                6.8               6.7                         0.452   
2929                7.8               7.8                         0.421   

      Schooling  
16         14.2  
17         14.2  
18         14.2  
22         12.2  
23         12.0  
...         ...  
2923       10.3  
2924       10.4  
2925        9.8  
2926       10.1  
2929        9.7  

[983 rows x 22 columns]

# Save the cleaned dataset

In [58]:
life.to_csv('CleanedLifeExpectancy.csv')

## Model

In [59]:
X=life.drop(columns='Lifeexpectancy')
y=life['Lifeexpectancy']

In [60]:
X

Country  Year      Status  AdultMortality  infantdeaths  Alcohol  \
0     Afghanistan  2015  Developing             263            62     0.01   
1     Afghanistan  2014  Developing             271            64     0.01   
2     Afghanistan  2013  Developing             268            66     0.01   
3     Afghanistan  2012  Developing             272            69     0.01   
4     Afghanistan  2011  Developing             275            71     0.01   
...           ...   ...         ...             ...           ...      ...   
2933     Zimbabwe  2004  Developing             723            27     4.36   
2934     Zimbabwe  2003  Developing             715            26     4.06   
2935     Zimbabwe  2002  Developing              73            25     4.43   
2936     Zimbabwe  2001  Developing             686            25     1.72   
2937     Zimbabwe  2000  Developing             665            24     1.68   

      percentageexpenditure  HepatitisB  Measles   BMI  ...  Polio  \
0                 71.279624          65     1154  19.1  ...      6   
1                 73.523582          62      492  18.6  ...     58   
2                 73.219243          64      430  18.1  ...     62   
3                 78.184215          67     2787  17.6  ...     67   
4                  7.097109          68     3013  17.2  ...     68   
...                     ...         ...      ...   ...  ...    ...   
2933               0.000000          68       31  27.1  ...     67   
2934               0.000000           7      998  26.7  ...      7   
2935               0.000000          73      304  26.3  ...     73   
2936               0.000000          76      529  25.9  ...     76   
2937               0.000000          79     1483  25.5  ...     78   

      Totalexpenditure  Diphtheria  HIV/AIDS  GDP  Population  \
0                 8.16          65       0.1  263    33736494   
1                 8.18          62       0.1  271      327582   
2                 8.13          64       0.1  268    31731688   
3                 8.52          67       0.1  272     3696958   
4                 7.87          68       0.1  275     2978599   
...                ...         ...       ...  ...         ...   
2933              7.13          65      33.6  723    12777511   
2934              6.52          68      36.7  715    12633897   
2935              6.53          71      39.8   73      125525   
2936              6.16          75      42.1  686    12366165   
2937              7.10          78      43.5  665    12222251   

      thinness1-19years  thinness5-9years  Incomecompositionofresources  \
0                  17.2              17.3                         0.479   
1                  17.5              17.5                         0.476   
2                  17.7              17.7                         0.470   
3                  17.9              18.0                         0.463   
4                  18.2              18.2                         0.454   
...                 ...               ...                           ...   
2933                9.4               9.4                         0.407   
2934                9.8               9.9                         0.418   
2935                1.2               1.3                         0.427   
2936                1.6               1.7                         0.427   
2937               11.0              11.2                         0.434   

      Schooling  
0          10.1  
1          10.0  
2           9.9  
3           9.8  
4           9.5  
...         ...  
2933        9.2  
2934        9.5  
2935       10.0  
2936        9.8  
2937        9.8  

[2938 rows x 21 columns]

In [61]:
y

0       65.0
1       59.9
2       59.9
3       59.5
4       59.2
        ... 
2933    44.3
2934    44.5
2935    44.8
2936    45.3
2937    46.0
Name: Lifeexpectancy, Length: 2938, dtype: float64

In [62]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline

In [63]:
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.2)

In [64]:
ohe=OneHotEncoder()
ohe.fit(X[['Country','Status']])

OneHotEncoder()

In [65]:
ohe.categories_

[array(['Afghanistan', 'Albania', 'Algeria', 'Angola',
        'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia',
        'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh',
        'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan',
        'Bolivia (Plurinational State of)', 'Bosnia and Herzegovina',
        'Botswana', 'Brazil', 'Brunei Darussalam', 'Bulgaria',
        'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cambodia', 'Cameroon',
        'Canada', 'Central African Republic', 'Chad', 'Chile', 'China',
        'Colombia', 'Comoros', 'Congo', 'Cook Islands', 'Costa Rica',
        'Croatia', 'Cuba', 'Cyprus', 'Czechia', "Côte d'Ivoire",
        "Democratic People's Republic of Korea",
        'Democratic Republic of the Congo', 'Denmark', 'Djibouti',
        'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt',
        'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia',
        'Ethiopia', 'Fiji', 'Finland', 'France', 'Gabon', 'Gambia',
        'Geo

In [66]:
column_trans=make_column_transformer((OneHotEncoder(categories=ohe.categories_),['Country','Status']),
                                     remainder='passthrough')

In [67]:
lr=LinearRegression()

In [68]:
pipe=make_pipeline(column_trans,lr)

In [69]:
pipe.fit(X_train,y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(categories=[array(['Afghanistan', 'Albania', 'Algeria', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia',
       'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh',
       'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhuta...
       'United Arab Emirates',
       'United Kingdom of Great Britain and Northern Ireland',
       'United Republic of Tanzania', 'United States of America',
       'Uruguay', 'Uzbekistan', 'Vanuatu',
       'Venezuela (Bolivarian Republic of)', 'Viet Nam', 'Yemen',
       'Zambia', 'Zimbabwe'], dtype=object),
                                                                            array(['Developed', 'Developing'], dtype=object)]),
                                                  ['Country', 'Status'])])),
                ('linearregression', LinearRegression())])

In [70]:
y_pred=pipe.predict(X_test)

In [71]:
r2_score(y_test,y_pred)

0.5942075933145716

In [72]:
scores=[]
for i in range(1000):
    X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.2, random_state=i)
    lr=LinearRegression()
    pipe=make_pipeline(column_trans,lr)
    pipe.fit(X_train,y_train)
    y_pred=pipe.predict(X_test)
    scores.append(r2_score(y_test,y_pred))
   

In [73]:
np.argmax(scores)

945

In [74]:
scores[np.argmax(scores)]

0.6685569151702632

In [75]:
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.2, random_state=np.argmax(scores))
lr=LinearRegression()
pipe=make_pipeline(column_trans,lr)
pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_test)
r2_score(y_test,y_pred)

0.6685569151702632

In [76]:
import pickle

In [77]:
pickle.dump(pipe,open('LinearRegressionModel.pkl','wb'))

# Prediction

In [78]:
pipe.predict(pd.DataFrame([['India',2022,'Developing',100,1000,3.4,56.6,80,9000,19.1,800,90,5.9,90,0.3,1500.1,1324517250,20.1,26.5,0.643,10.1]],columns=['Country','Year','Status','AdultMortality','infantdeaths','Alcohol','percentageexpenditure','HepatitisB','Measles','BMI','under-fivedeaths','Polio','Totalexpenditure','Diphtheria','HIV/AIDS','GDP','Population','thinness1-19years','thinness5-9years','Incomecompositionofresources','Schooling']))

array([48.55879138])